src/_pytest/subtests.py

```diff
@@ -0,0 +1,431 @@
+    ) -> None:
+        check_ispytest(_ispytest)
+        self._ihook = ihook
+        self._suspend_capture_ctx = suspend_capture_ctx
+        self._request = request
+
+    def test(
+        self,
+        msg: str | None = None,
+        **kwargs: Any,
+    ) -> _SubTestContextManager:
+        """
+        Context manager for subtests, capturing exceptions raised inside the subtest scope and
+        reporting assertion failures and errors individually.
+
+        Usage
+        -----
+
+        .. code-block:: python
+
+            def test(subtests):
+                for i in range(5):
+                    with subtests.test(msg="custom message", i=i):
+                        assert i % 2 == 0
+
+        :param msg:
+            If given, the message will be shown in the test report in case of subtest failure.
+
+        :param kwargs:
+            Arbitrary values that are also added to the subtest report.
+        """
+        return _SubTestContextManager(
+            self._ihook,
+            msg,
+            kwargs,
+            request=self._request,
+            suspend_capture_ctx=self._suspend_capture_ctx,
+            reporter=self._request.config.pluginmanager.getplugin(
+                SubtestsReporterPlugin.NAME
+            ),
+        )
+
+
+@dataclasses.dataclass
+class _SubTestContextManager:
+    """
+    Context manager for subtests, capturing exceptions raised inside the subtest scope and handling
+    them through the pytest machinery.
+    """
+
+    # Note: initially the logic for this context manager was implemented directly
+    # in Subtests.test() as a @contextmanager, however, it is not possible to control the output fully when
+    # exiting from it due to an exception when in `--exitfirst` mode, so this was refactored into an
+    # explicit context manager class (pytest-dev/pytest-subtests#134).
+
+    ihook: pluggy.HookRelay
+    msg: str | None
+    kwargs: dict[str, Any]
+    suspend_capture_ctx: Callable[[], AbstractContextManager[None]]
+    request: SubRequest
+    reporter: SubtestsReporterPlugin
+
+    def __enter__(self) -> None:
+        __tracebackhide__ = True
+
+        self._start = time.time()
+        self._precise_start = time.perf_counter()
+        self._exc_info = None
+
+        self._exit_stack = ExitStack()
+        self._captured_output = self._exit_stack.enter_context(
+            capturing_output(self.request)
+        )
+        self._captured_logs = self._exit_stack.enter_context(
+            capturing_logs(self.request)
+        )
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> bool:
+        __tracebackhide__ = True
+        try:
+            if exc_val is not None:
+                exc_info = ExceptionInfo.from_exception(exc_val)
+            else:
+                exc_info = None
+        finally:
+            self._exit_stack.close()
```

Comment:
I don't think `ExceptionInfo.from_exception` is something that can fail, so is there a reason for wrapping in try/finally?